arch: xtensa: save/restore the HiFi (AudioEngine) registers

peter-mitsis opened the pull request:
This is an initial set of commits designed to add support to unconditionally save/restore the AudioEngine registers from the HiFi coprocessor.

Proper tests have yet to be written, although some initial testing has been done ...
  1. Disassembled code indicates that the code is generated when the feature is enabled
  2. Earlier errors during development showed that this code is getting executed
  3. Manually enabled tests have been passing.

It is recognised that this testing is obviously incomplete and that further changes are expected as a result of the feedback, community knowledge and coordination with other developers that are touching upon related areas (such as #67513).

(For now, please ignore the 1st two commits dealing with rsr.XXX and wsr.XXX instructions. Those commits have their own PR, but are only included here because they have not finished review and thus have not been merged.)

**Expected Objections**

1. _When the feature is enabled, the AE registers are unconditionally saved restored, which is definitely going to have some performance implications._ This is an early draft aiming for a minimal viable product (MVP). Enabling conditional saving/restoring of these registers is going to require further work, and is not in the planned scope of this PR, but rather a future one.
2. _Kconfig options and conditions used to include/exclude this feature._ Admittedly, I am not happy yet with the set of conditions I have initially used to include/exclude this feature. These should be extended modified to use some sort of Kconfig such as CONFIG_HIFIx. I am aware of #67513, but that has not yet been merged, and some coordination will be required. It is acknowledged that changes here will be necessary.
3. _Why not generalise this for all potential xtensa co-processors? (After all, the HAL specifies the sizes and alignments for up to 8 co-processors.)_ This is not off the table, but my primary concern at this point in time is enabling it for the HiFi4.

**Open Questions**
1. For a future per-thread enablement, would a Zephyr developer use the existing K_DSP_REGS with k_thread_create()? Or should it use something like a new K_HIFI_REGS?
2. Is there another co-processor (CP) aside from the HiFiX that might be CP1. I found a reference in the HiFi documentation saying that it is always CP1, but did not find anything that indicated that nothing else could be CP1 either.


I am sure that there will be other items. :)

peter-mitsis committed:
arch: xtensa: Add space for HiFi registers

Updates the xtensa_irq_base_save_area structure to include space
for saving/restoring the HiFi AudioEngine registers used by CP1.

The starting address of these HiFi AE registers also needs to be
referenced from assembly, so it is added to the set of symbols
symbols for which we need an offset to be auto-generated.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
arch: xtensa: save/restore HiFi AudioEngine regs

Adds the necessary code required to unconditionally save/restore the
HiFi AE registers. The macros xchal_cp1_load and xchal_cp1_store
are defined in the Xtensa HAL.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

btian1 commented on arch/xtensa/Kconfig:
```diff
@@ -93,6 +93,24 @@ config XTENSA_NUM_SPIN_RELAX_NOPS
+if CPU_HAS_DSP
+menu "DSP Options"
+
+config DSP_SHARING
```
There is no specific HIFI DSP classification, @peter-mitsis , so there is no difference for HIFI3/4/5 for context save/restore?

singalsu replied:
I wonder if CPU_HAS_XTENSA_HIFI would be a better kconfig name since this is about storing hifi registers?

peter-mitsis replied:
Your idea seems to align well with  #67513--happy to move in that direction. :)

btian1 commented on arch/xtensa/core/offsets/offsets.c:
```diff
@@ -61,6 +61,11 @@ GEN_OFFSET_SYM(_xtensa_irq_bsa_t, fpu14);
 
+#if (XCHAL_CP_MASK & (1 << XCHAL_CP_ID_AUDIOENGINELX)) && \
+	defined(CONFIG_CPU_HAS_DSP) && defined(CONFIG_DSP_SHARING)
+GEN_OFFSET_SYM(_xtensa_irq_bsa_t, ae_regs);
```
any difference between different DSP architecture? @singalsu

singalsu replied:
There's difference but I wonder if the aligned address and size macros used here cover those. I don't know.

peter-mitsis replied:
The aligned and size macros are defined in the HAL and should cover those differences.

peter-mitsis committed:
tests: Add test to verify HiFi save/restore

Adds a test to verify the saving and restoration of the HiFi
AudioEngine registers. Twice as many threads as there are CPUs
are created and each thread sets its HiFi registers with a
unique set of values. If after a context switch, the HiFi AE
registers contain any unexpected data, we then know that the
registers were either not saved or restored properly.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
arch: xtensa: Add HiFi Kconfig options

Adds HiFi Kconfig options to xtensa

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

marc-hb commented on arch/xtensa/include/xtensa_asm2_s.h:
```diff
@@ -9,6 +9,8 @@
 #include <zsr.h>
 #include "xtensa_asm2_context.h"
+#include <xtensa/config/tie.h>
+#include <xtensa/config/tie-asm.h>
```
One of the (admittedly unclear and debatable) requirement behind #67513 was NOT to include .h files from the HAL and NOT get `XCHAL_*_HIFIn` macros from it.  Duplicate the HAL macros in Kconfig instead.

If you're including files from the HAL, then do you really need #67513 as is? Or a much smaller and more dynamic version of it?

ceolin commented on arch/xtensa/include/xtensa_asm2_s.h:
```diff
@@ -184,6 +186,20 @@
+	and a2, a2, a3
+
+	xchal_cp1_store a2 a0 a3 a3 a3  /* Only A0 and A2 are used by macro */
+#endif
```
This macro says that a2/a3 won't be modified at exit. I didn't see any place assuming it, but it is documented and need to be confirmed and then, at least change the documentation.

ceolin commented on arch/xtensa/include/xtensa_asm2_context.h:
```diff
@@ -106,6 +106,18 @@ struct xtensa_irq_base_save_area {
+	 * at run-time.
+	 */
+	uint8_t  ae_regs[XCHAL_CP1_SA_SIZE + XCHAL_CP1_SA_ALIGN];
+#endif
```
Any particular reason why this is located here ? Can it end up leaving a hole in the struct depending on XCHAL_CP1_SA_SIZE + XCHAL_CP1_SA_ALIGN ?

peter-mitsis replied:
Assuming the first question refers to why it was placed in the `struct xtensa_irq_base_save_area` -- it seemed like a reasonable place to store the AudioEngine HiFi registers as the floating point registers were already here.

As to the potential hole ... there is definitely going to be some unused space. When the HiFi3 or HiFi4 is enabled, XCHAL_CP1_SA_ALIGN evaluates to 8. However, from what I can tell, `struct xtensa_irq_base_save_area` is not guaranteed to be aligned to an 8-byte boundary (see its use in `struct xtensa_irq_stack_frame_aX` where X = {3, 7, 11, 15}). To work around this, the alignment has been added to the size to provide extra space so that we can force the necessary alignment at runtime.

andyross replied:
Sorry, repeating from elsewhere:  Indeed, the BSA isn't aligned in any special way on entry, but there's no reason that couldn't be the case.  Just put the ae_regs[] array at an aligned spot within this struct, and align the stack before pushing the registers.

peter-mitsis replied:
For what it is worth, initial simple attempts to just align the structure via the` __align()` macro are not sufficient (resulting in fatal exceptions). This suggests that there is assembly code that carves out space for the BSA that results in an unaligned BSA. Subsequent searches for a generated offset that would provide further evidence has turned up `___xtensa_irq_bsa_t_SIZEOF`, which is indeed used in the assembly files to carve space (or at least adjust the pointers). I am currently digging into what exactly would need to force the correct alignment when those pointers are updated.

andyross commented on arch/xtensa/include/xtensa_asm2_s.h:
```diff
@@ -160,20 +160,20 @@
  */
 .macro ODD_REG_SAVE
-	rsr.SAR a0
+	rsr.sar a0
```
Which header is colliding?  Our own symbols should all be namespaced.  This sounds to me like we're pulling in some HAL code that we really don't want...

peter-mitsis replied:
modules/hal/xtensa/zephyr/soc/[board]/xtensa/config/specreg.h

andyross commented on arch/xtensa/include/xtensa_asm2_s.h:
```diff
@@ -371,7 +371,7 @@ _xstack_call0_\@:
 	mov a1, a2
 	rsr.ZSR_EPS a2
-	wsr.PS a2
+	wsr.ps a2
```
Why are "rsr" and "wsr" two different patches? :)

peter-mitsis replied:
That wound up getting handled separately. It was an artefact of my workflow--no real reason why could not have been squashed.

andyross commented on arch/xtensa/Kconfig:
```diff
@@ -93,6 +93,24 @@ config XTENSA_NUM_SPIN_RELAX_NOPS
+	help
+	  This option enables preservation of the hardware DSP registers
+	  across context switches to allow multiple threads to perform
+	  concurrent DSP operations.
```
So, I get why this is called "SHARING" by symmetry with the existing FPU features.  But with Xtensa, existing toolchains will generate HiFi instructions that use the AE registers whether or not the app wants them to[1].  There's really no feasible situation where we'd want this disabled (or rather, it would be feasible, but would require apps directly control compiler flags and probably enforce their own ABIs).

Suggest removing the kconfig here and just keying off of XCHAL_HAVE_xxx from core-isa.h

[1] gcc doesn't, and in general xt-clang won't for scalar code, but if you look at the generated assembly for anything that looks like array processing you'll see HiFi bundles in the output.

peter-mitsis replied:
I still like the "SHARING" for purposes of symmetry. However, you raise a very interesting point with regards to toolchain behaviour. I had originally set "SHARING" to 'n' so that we would not suffer stack size nor context switch runtime penalties. However, it might be worthwhile to have the default set according to the board or soc.

peter-mitsis replied:
I've been doing some more reading on this, and it looks like xt-clang may automatically generate HiFi code if any of the following are met ...

1. Automatic vectorization is enabled (via `O3, -LNO:simd`, or `-fvectorize`, or )
2.  Or if the `-mcoproc` option is used.

Near as I can tell, these are not enabled by default in Zephyr. Furthermore, I do not yet see any options for them to be enabled in Zephyr. Assuming the both the documentation and my interpretation are correct, then it may be reasonable _at this point in time_ to have the project be responsible for the _sharing_ and compilation flags.

andyross commented on arch/xtensa/include/xtensa_asm2_context.h:
```diff
@@ -106,6 +106,18 @@ struct xtensa_irq_base_save_area {
 	uintptr_t fpu15;
 #endif
 
+#if (XCHAL_CP_MASK & (1 << XCHAL_CP_ID_AUDIOENGINELX)) && \
```
Can you key this off of core-isa.h values and not the derived junk from the Xtensa HAL?  XCHAL_HAVE_HIFI{3,4,5}, etc...

andyross commented on arch/xtensa/core/xtensa_asm2_util.S:
```diff
@@ -160,6 +160,19 @@ _restore_context:
+	addi a2, a2, (XCHAL_CP1_SA_ALIGN - 1)
+	and a2, a2, a3
+
+	xchal_cp1_load a2 a0 a3 a3 a3    /* Only A0 and A2 are used by macro */
```
As mentioned, really really not a fan of seeing HAL code here in the core entry code, especially since we all agree we're going to throw out this technique in the future anyway and move to a lazy-evaluated version based on the CP enable trap.

If we really have to merge this version first, can we move it to a separate .S file where we can isolate the HAL pollution and call it as a subroutine (or just a CALL0 I guess, since there's no stack needed and the registers get clobbered anyway)?

peter-mitsis replied:
Isolated.

andyross reviewed (commented):
Sorry, completely missed this when it came by earlier.  Some notes.

Really the big one is that the use of the HAL assembly routines here is IMHO a mistake we're going to need to unwind, and frankly probably more complicated than doing this the "right way" via the CP1 exception anyway.   Note that in the exception case, you end up in the handler in C code and can do the spills with a ton more simplicity and flexibility (i.e. no need to muck with getting it on the stack in an aligned form, just dump it in the thread struct).

andyross commented:
Also, would be nice to see some thought given to unifying this code and the pre-existing Xtensa FPU handling.  Cadence (to their credit) actually have a pleasingly orthogonal coprocessor architecture and it would be sort of upsetting for these not to share a backend.  Also note that Intel TGL/ADL/RPL hardware[1] has a HiFi3 unit without floating point, and thus need to use the scalar FPU instead.

[1] There's a weird irony that I have to care more about "old" Intel hardware than Intel itself does, I recognize.  Zephyr isn't shipped yet on these devices, but it does run and there's still hope to get a unified tree.

andyross commented:
Oh, and one more: obviously this is going to add 128+ bytes to every stack on the system, which is likely going to break a bunch of Zephyr tests (though probably not SOF itself which uses bigger stacks).  Obviously none of the CI platforms are going to be any help with HiFi, so I guess the only option is hand-running twister against MTL?  Would be helpful to report success.

andyross commented on arch/xtensa/include/xtensa_asm2_s.h:
```diff
@@ -184,6 +186,20 @@
+	addi a2, a1, ___xtensa_irq_bsa_t_ae_regs_OFFSET
+	movi a3, ~(XCHAL_CP1_SA_ALIGN - 1)
+	addi a2, a2, (XCHAL_CP1_SA_ALIGN - 1)
+	and a2, a2, a3
```
This seems like a clumsy way to do the alignment.  Why not just whack off the bottom three bits of a1 above at the start of entry code to guarantee that everything is correctly aligned?

peter-mitsis replied:
Done.

andyross commented:
Just to pull this down from above and reply while it's in my head:

> Is there another co-processor (CP) aside from the HiFiX that might be CP1.

Almost certainly not, but CP1 might be any of HiFi3/4/5 with or without variants that change the saved state.  So there's not much that can be won by assumption anyway.  Really what I think we want here is an architecture where every coprocessor (note that we already have code for the scalar FPU in tree!) gets to declare itself and its type/parametrization in DTS somewhere, and that then gets handled by a generic save/restore framework at the level of the generic CP access exceptions.

dcpleung commented on include/zephyr/kernel.h:
```diff
@@ -246,6 +246,18 @@ void k_thread_foreach_unlocked(
+ * these registers when scheduling the thread. No effect if
+ * @kconfig{CONFIG_XTENSA_HIFI_SHARING} is not enabled.
+ */
+#define K_HIFI_IDX 7
+#define K_HIFI_REGS (BIT(K_HIFI_IDX))
```
Don't think it is necessary... as the code is saving/restoring as long as kconfig is enabled. There is no per thread settings.

peter-mitsis replied:
Nice catch. I was originally going to try for a per-thread setting but wound up not since there did not seem to be a clean way I recognised to get the pointer to the thread.

 Will remove.

peter-mitsis commented:
Rebased.

nashif commented on tests/arch/xtensa/save_restore_hifi/testcase.yaml:
```diff
@@ -0,0 +1,6 @@
+tests:
+  xtensa.save_restore_hifi.mtl:
+    platform_allow: intel_adsp_ace15_mtpm
```
you need to use the new platform name, and please use a yaml list
